backend/cmm_builtins.ml

```diff
@@ -195,6 +195,14 @@ let bigstring_atomic_add size (arg1, arg2, arg3) dbg =
 let bigstring_atomic_sub size (arg1, arg2, arg3) dbg =
   bigstring_atomic_add size (arg1, arg2, neg_int arg3 dbg) dbg
 
+(* Assumes unboxed float32 *)
+let rec const_float32_args n args name =
```

Comment:
(This is the fourth variant of this function;
it might be worth factoring out.)

Reply:
Hmm, I tried combining them but I don't think it's worthwhile, since they have different types/constructors/error messages

Reply:
I agree a patch such as the one below is not
particularly elegant, but won't we have more
copy-paste when we introduce additional
small numbers?

```diff
--- a/backend/cmm_builtins.ml
+++ b/backend/cmm_builtins.ml
@@ -195,38 +195,28 @@ let bigstring_atomic_add size (arg1, arg2, arg3) dbg =
 let bigstring_atomic_sub size (arg1, arg2, arg3) dbg =
   bigstring_atomic_add size (arg1, arg2, neg_int arg3 dbg) dbg

-(* Assumes unboxed float32 *)
-let rec const_float32_args n args name =
+let rec const_args_gen ~extract ~type_name n args name =
   match n, args with
   | 0, [] -> []
-  | n, Cconst_float32 (f, _) :: args ->
-    f :: const_float32_args (n - 1) args name
-  | _ -> bad_immediate "Did not find constant float32 arguments for %s" name
+  | _, [] -> assert false (* TODO *)
+  | n, arg :: args ->
+    begin match extract arg with
+    | Some value -> value :: const_args_gen ~extract ~type_name (n - 1) args name
+    | None -> bad_immediate "Did not find constant %s arguments for %s" type_name name
+    end
+
+(* Assumes unboxed float32 *)
+let const_float32_args = const_args_gen ~extract:(function Cconst_float32 (f, _) -> Some f | _ -> None) ~type_name:"float32"

 (* Assumes unboxed float64 *)
-let rec const_float_args n args name =
-  match n, args with
-  | 0, [] -> []
-  | n, Cconst_float (f, _) :: args -> f :: const_float_args (n - 1) args name
-  | _ -> bad_immediate "Did not find constant float arguments for %s" name
+let const_float_args = const_args_gen ~extract:(function Cconst_float (f, _) -> Some f | _ -> None) ~type_name:"float"

 (* Assumes untagged int or unboxed int32, always representable by int63 *)
-let rec const_int_args n args name =
-  match n, args with
-  | 0, [] -> []
-  | n, Cconst_int (i, _) :: args -> i :: const_int_args (n - 1) args name
-  | _ -> bad_immediate "Did not find constant int arguments for %s" name
+let const_int_args = const_args_gen ~extract:(function Cconst_int (i, _) -> Some i | _ -> None) ~type_name:"int"

 (* Assumes unboxed int64: no tag, comes as Cconst_int when representable by
    int63, otherwise we get Cconst_natint *)
-let rec const_int64_args n args name =
-  match n, args with
-  | 0, [] -> []
-  | n, Cconst_int (i, _) :: args ->
-    Int64.of_int i :: const_int64_args (n - 1) args name
-  | n, Cconst_natint (i, _) :: args ->
-    Int64.of_nativeint i :: const_int64_args (n - 1) args name
-  | _ -> bad_immediate "Did not find constant int64 arguments for %s" name
+let rec const_int64_args = const_args_gen ~extract:(function Cconst_int (i, _) -> Some (Int64.of_int i) | Cconst_natint (i, _) -> Some (Int64.of_nativeint i) | _ -> None) ~type_name:"int64"

 let int64_of_int8 i =
   (* CR mslater: (SIMD) replace once we have unboxed int8 *)

```

Reply:
Yeah, that's true - updated to your suggestion